src/Neo/SmartContract/Native/NativeContract.cs

```diff
@@ -107,28 +113,58 @@ public abstract class NativeContract
-        public ContractManifest Manifest { get; }
-
         /// <summary>
         /// Initializes a new instance of the <see cref="NativeContract"/> class.
         /// </summary>
         protected NativeContract()
         {
-            List<ContractMethodMetadata> descriptors = new();
+            this.Hash = Helper.GetContractHash(UInt160.Zero, 0, Name);
+
+            // Reflection to get the methods
+
+            List<ContractMethodMetadata> listMethods = new();
             foreach (MemberInfo member in GetType().GetMembers(BindingFlags.Instance | BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public))
             {
                 ContractMethodAttribute attribute = member.GetCustomAttribute<ContractMethodAttribute>();
                 if (attribute is null) continue;
-                descriptors.Add(new ContractMethodMetadata(member, attribute));
+                listMethods.Add(new ContractMethodMetadata(member, attribute));
             }
-            descriptors = descriptors.OrderBy(p => p.Name, StringComparer.Ordinal).ThenBy(p => p.Parameters.Length).ToList();
+            methodDescriptors = listMethods.OrderBy(p => p.Name, StringComparer.Ordinal).ThenBy(p => p.Parameters.Length).ToList().AsReadOnly();
+
+            // Reflection to get the events
+            eventsDescriptors =
+                GetType().GetConstructor(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public, Array.Empty<Type>())?.
+                GetCustomAttributes<ContractEventAttribute>().
+                OrderBy(p => p.Order).ToList().AsReadOnly();
+
+            // Calculate the initializations forks
+            listenHardforks =
+                methodDescriptors.Select(u => u.ActiveIn)
+                .Concat(eventsDescriptors.Select(u => u.ActiveIn))
+                .Concat(new Hardfork?[] { ActiveIn })
+                .Where(u => u is not null)
+                .OrderBy(u => (byte)u)
+                .Cast<Hardfork>().ToImmutableHashSet();
+
+            contractsList.Add(this);
+            contractsDictionary.Add(Hash, this);
+        }
+
+        /// <summary>
+        /// The allowed methods and his offsets.
+        /// </summary>
+        /// <param name="settings">The <see cref="ProtocolSettings"/> where the HardForks are configured.</param>
+        /// <param name="index">Block index</param>
+        /// <returns>The <see cref="NefMethods"/>.</returns>
+        private NefMethods GetAllowedMethods(ProtocolSettings settings, uint index)
+        {
+            Dictionary<int, ContractMethodMetadata> methods = new();
+
+            // Reflection to get the ContractMethods
             byte[] script;
             using (ScriptBuilder sb = new())
             {
-                foreach (ContractMethodMetadata method in descriptors)
+                foreach (ContractMethodMetadata method in methodDescriptors.Where(u => u.ActiveIn is null || settings.IsHardforkEnabled(u.ActiveIn.Value, index)))
```

Comment:
Looks good to have the list of the filtered methods by height. I would suggest that as well.

Is this checked and updated every OnPersist?

Reply:
I think that Consensus should emit a proposed Hardfork with the index and other agree. Then, client won't need to reestart.

Reply:
clients need the new logic for this hardfork, the need to update the binaries

Reply:
In this case, @shargon,
If CNs update their clients and start to produce new blocks.
A client will sync it?
Perhaps the client should stuck at least, no? Otherwise it might be complex to revert.

Reply:
At the beginning I thought a payload would be sent enabling it to clients.